[NEW PR] Dedicated WandbLogger for MMDetection

ayulockin opened the pull request:
This PR is opened to sign the CLA agreement. The previous PR #7413 is closed in favor of this. You can check that PR for initial discussion. 

The details of the feature can be found in this issue #7391.

ZwwWayne commented on mmdet/core/evaluation/eval_hooks.py:
```diff
@@ -114,15 +114,15 @@ def _do_evaluate(self, runner):
 
         from mmdet.apis import multi_gpu_test
-        results = multi_gpu_test(
+        self.results = multi_gpu_test(
```
Is this modification necessary?

ayulockin replied:
Hey @ZwwWayne this modification is necessary so that MMDetWandbHook can reuse the result.

ayulockin commented:
Hey all, the `codecov` workflow is failing. I guess it's because a unit test is not covering the added lines of code. I can write the unit test but would love to get some direction on this. Can you point me to a document/link or example tests based on which I can write the test?

ayulockin commented:
Hey @ZwwWayne, @RangiLyu, @hhaAndroid, @jbwang1997 would love to get some feedback/review on this PR. Apologies to ping again. Just wanted to confirm that there's nothing remaining from my side on this. Thanks!

jbwang1997 commented:
Sorry for the later reply. We are currently busy with the vision release. We have planned to merge this pr in as soon as the next vision and I leave some comments below. Thank you so much for your valuable contribution.

jbwang1997 commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,455 @@
+    @master_only
+    def before_run(self, runner):
+        super(MMDetWandbHook, self).before_run(runner)
+        self.cfg = self.wandb.config
+        # Check if configuration is passed to wandb.
+        if len(dict(self.cfg)) == 0:
+            warnings.warn(
+                'To log mmdetection Config, '
+                'pass it to init_kwargs of MMDetWandbHook.', UserWarning)
```
After consideration. It's no good to modify the inputs of runner. I consider if we can obtain the config in the work_dir.

ayulockin replied:
That's a great idea. Let me try to grab the config from the `work_dir`.

hhaAndroid replied:
We have saved the configuration so you can read it directly.

ayulockin replied:
Hey @hhaAndroid can you point me to a code where the configuration is saved. I couldn't find any saved configuration in the `work_dir`.

hhaAndroid replied:
Although it is stored, but it is not easy to read, I am trying to figure out a way

hhaAndroid replied:
```python
dict(type='MMDetWandbHook',
        init_kwargs={
                'project': 'atss_r50_fpn_1x_coco'
        },
       interval=10,
        log_checkpoint=True,
        log_checkpoint_metadata=True,
        num_eval_images=100,
        saved_config_name='{{ fileBasename }}')


# Load config.json file from work_dir
load_config_path = osp.join(runner.work_dir, self.saved_config_name)
if osp.exists(load_config_path):
      self.wandb.save(load_config_path)
else:
     warnings.warn('`saved_config_name` does not exist, '
              'skip saving the config file directly', UserWarning)
```

ayulockin replied:
Hey I am not clear about `args.config`. 

Also when I use '{{ fileBasename }}' gives `/home/ayut/MMDet/mmdetection/work_dirs/faster_rcnn_r50_fpn_1x_coco/default_runtime.py` which I feel is just a small fraction of the complete configuration. 

The current solution that's in the PR is capable of logging the complete config.

ayulockin replied:
Hey @hhaAndroid, apologies to unresolve this conversation but just wanted to be sure if my method to log config is acceptable?

I was not able to make your solution work as explained above.

jbwang1997 commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,455 @@
+        table_idxs = self.data_table_ref.get_index()
+        assert len(table_idxs) == self.num_eval_images
+
+        for ndx in table_idxs:
+            result = results[ndx]
+            assert len(result) == len(self.class_id_to_label)
```
mmdetection also support instance segmentation model like mask rcnn. Their outputs is an tuple which is an combination of bbox results and segm results. Could you also support instance segmentation outputs in this pr?

ayulockin replied:
I can work on supporting instance segmentation as well. But can you point me to an example/tutorial that used MMDetection for the segmentation task?

PS: W&B lacks good support for "instance" segmentation at the moment but I can try making a workaround.

ayulockin replied:
Also does MMDetection support "semantic" segmentation?

jbwang1997 replied:
You can refer to the [show_result](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/detectors/base.py#L271) function for the format of mask results. If you have any questions, you can directly point them out here.

jbwang1997 replied:
Semantic segmentation is not supported.

ayulockin replied:
As mentioned above W&B lacks first-class support for "instance" segmentation and for this reason I will be adding this feature once W&B build support for this. 

For now, the `results` format is handled properly. Thanks for pointing this out.

hhaAndroid commented:
@ayulockin I will update with you.

hhaAndroid commented:
@ayulockin Hello
1. mmcv's wandbhook already supports log eval metrics, can your PR delete this part of the code?
2. I see that mask visualization is not supported yet.
3.

ayulockin commented:
> 1. mmcv's wandbhook already supports log eval metrics, can your PR delete this part of the code?

Yes I can remove this part. 

> 2. I see that mask visualization is not supported yet.

Yes, wandb doesn't support instance segmentation as of yet.

hhaAndroid commented:
> 

I see it's supported https://docs.wandb.ai/ref/python/data-types/imagemask. Each bbox comes with a mask_data to complete instance segmentation visualization

julled commented:
@ayulockin i really like this PR, thx for implementing all the valuable functions!

I am not a super duper expert in mlflow but from what i see from your code it might be that the functions you introduce in your PR are with minor changes also nicely usable for a smiliar mlflow hook.

Do you there might be some generic functions to be put in a kind of "common" library to use e.g. for a similar mlflow hook extension?

ayulockin commented:
> @ayulockin i really like this PR, thx for implementing all the valuable functions!
> 
> I am not a super duper expert in mlflow but from what i see from your code it might be that the functions you introduce in your PR are with minor changes also nicely usable for a smiliar mlflow hook.
> 
> Do you there might be some generic functions to be put in a kind of "common" library to use e.g. for a similar mlflow hook extension?

Hey @julled, glad you liked the work. 

I have not used MLFlow myself so can't say if we can abstract away some of the class methods as a utility that can be used for the MLFlow hook. The `MMDetWandbHook` introduced in this PR inherits `WandbLoggerHook`. If you wanna implement an extended hook for MLFlow there is a `MlflowLoggerHook` that you can inherit.

Also, the hook that this PR is introducing is adding functionalities that are specific to W&B.

I hope this helps.

ayulockin commented:
> I see it's supported https://docs.wandb.ai/ref/python/data-types/imagemask. Each bbox comes with a mask_data to complete instance segmentation visualization

Hey @hhaAndroid, yes logging mask_data is supported by W&B but the issue is that instance segmentation visualization is not supported by the UI properly. If the mask data is for semantic segmentation then the UI can render it nicely. Since MMDetection doesn't support semantic segmentation, I think the feature will not be very useful now. 

I can add the feature once instance segmentation visualization is supported by W&B. Thoughts?

hhaAndroid commented:
> 

It doesn't matter, you can display the mask and bbox together to see the effect. If the effect is better, it can be supported

hhaAndroid commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,443 @@
+            bool: Returns True, if the checkpoint has the best metric score.
+        """
+        keys = list(metadata.keys())
+        map_metrics = [key for key in keys if 'mAP' in key]
```
This function cannot handle panoptic segmentation, and there is no generality. My suggestion is that it does not need to support this function.

ayulockin replied:
Is there a way to programmatically find the evaluation metrics? Or is there a set of evaluation metrics upon which we can determine which checkpoint is the best one? 

I personally find it useful that in the UI, I have the information about the best checkpoint.

hhaAndroid replied:
There is no particularly good way for the time being, if we must support it, we can only hard code.

ayulockin replied:
I have removed this for now.

hhaAndroid commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,443 @@
+                    runner.model, self.val_dataloader, show=False)
+            elif isinstance(self.eval_hook, DistEvalHook):
+                from mmdet.apis import multi_gpu_test
+                results = multi_gpu_test(
```
It is unreasonable to need an evaluation in order to get meta. If there is no other way, I think it's better not to support this feature, what do you think?

ayulockin replied:
Are you are saying that it's not reasonable to run an extra evaluation step just to get metadata for checkpoint? 

To be honest I was sceptical as well since it will be a computational overhead. This is more of an issue especially if the dataset is huge. However, this extra evaluation will happen at each checkpoint interval and ideally for longer training checkpointing is done at reasonable intervals. 

I think it's best to use modulo operator with checkpoint interval and eval interval. If the modulo is zero then we have the `results` and need not evaluate again. Bottomline is it's best if we remove this.

hhaAndroid replied:
Yes, In principle,  We can't have an additional evaluation process. If there is no current evaluation, consider skipping it.

ayulockin commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,443 @@
+            self.num_eval_images = 0
+            self.log_checkpoint_metadata = False
+            self.log_eval_metrics = False
+            warnings.warn(
```
Just to be clear you think it's best if we remove this warning? The try-except block is fine, right?

hhaAndroid replied:
we can improve the warning to make it more friendly

ayulockin commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,443 @@
+        # If CheckpointHook is not available turn off log_checkpoint.
+        if getattr(self, 'ckpt_hook', None) is None:
+            self.log_checkpoint = False
+            warnings.warn('To use log_checkpoint turn use '
```
I think there are typos in the warning message. It's not gramatically correct. Do you think correcting for the grammar will make it friendly. 

Or do you want to remove the warning message?

hhaAndroid replied:
we can improve the warning to make it more friendly

ayulockin commented:
Hey all, thank you for the patience with this one. I have successfully implemented ground truth segmentation mask logging, I just have to make some tweaks to log prediction segmentation masks. I will be able to make the commit with all the changes tomorrow. Thanks.

ayulockin commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,492 @@
+                 init_kwargs=None,
+                 interval=50,
+                 log_checkpoint=False,
+                 log_checkpoint_metadata=False,
```
I think `log_checkpoint_metadata` is not necessary anymore. I will remove it.

hhaAndroid replied:
OK

hhaAndroid commented:
> 

Very Good

hhaAndroid commented:
@ayulockin Please resolve conflicts.

ayulockin commented:
Hey, @hhaAndroid I have resolved the conflict issues and removed extra arg. Would love to know if anything else is required from my end. :)

hhaAndroid commented:
> 

OK. We will deal with it asap.

hhaAndroid commented on mmdet/apis/train.py:
```diff
@@ -125,6 +125,10 @@ def train_detector(model,
     logger = get_root_logger(log_level=cfg.log_level)
 
+    # write config to work_dir
+    save_config_path = os.path.join(cfg.work_dir, 'config.py')
```
delete. there are better implementations

hhaAndroid commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,482 @@
+        super(MMDetWandbHook, self).before_run(runner)
+
+        # Load config.json file from work_dir
+        load_config_path = osp.join(runner.work_dir, 'config.py')
```
```python
        src_cfg_path = osp.join(runner.work_dir, runner.meta['exp_name'])
        if osp.exists(src_cfg_path):
               self.wandb.save(src_cfg_path)
```

ayulockin replied:
Unfortunately for `runner.meta['exp_name']` I am getting `TypeError: 'NoneType' object is not subscriptable`.

hhaAndroid replied:
Hey, it's impossible, I've already run it without a problem.

hhaAndroid replied:
Please confirm it, thank you.

hhaAndroid commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,482 @@
+            warnings.warn(
+                'LoadImageFromFile is required to add images '
+                'to W&B Tables.', UserWarning)
+            self.log_evaluation = False
```
```python
self.log_evaluation = False
return

hhaAndroid commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,482 @@
+
+        img_prefix = self.val_dataset.img_prefix
+
+        for idx in range(self.num_eval_images):
```
replace:
```python

eval_image_indexs = np.arange(self.num_eval_images)
np.random.shuffle(eval_image_indexs)
for idx in eval_image_indexs:
      ....

ayulockin replied:
This is a better implementation. Thanks. :)

hhaAndroid commented on mmdet/core/evaluation/eval_hooks.py:
```diff
@@ -53,9 +53,10 @@ def _do_evaluate(self, runner):
 
         from mmdet.apis import single_gpu_test
-        results = single_gpu_test(runner.model, self.dataloader, show=False)
+        self.results = single_gpu_test(
```
Add a comment explaining why I changed it to self.

hhaAndroid commented:
@ayulockin 

1. Please add `Wandbhook` to [it](https://github.com/open-mmlab/mmdetection/pull/7810) once it is complete.
2. If possible, please provide a new configuration `configs/mask_rcnn/mask_rcnn_r50_fpn_1x_wandb_coco.py` as an example, then run this configuration. We will use the link you post as a promotional visualization. Thank you.

hhaAndroid commented:
@ayulockin Thank you very much for your patience to modify. Look good to me.

hhaAndroid commented on mmdet/core/evaluation/eval_hooks.py:
```diff
@@ -114,15 +115,15 @@ def _do_evaluate(self, runner):
 
         from mmdet.apis import multi_gpu_test
-        results = multi_gpu_test(
+        self.results = multi_gpu_test(
```
It may be necessary to confirm whether the memory will increase after changing to the self attribute. Because result will be resident in memory

ayulockin replied:
I checked the current implementation and without using the same (and removed self from self.results) - there is no change in the memory requirement while training is going on. I didn't notice any increase in the memory because of the modification to eval_hooks.py. :)

ayulockin commented:
Hey @hhaAndroid I have made the changes that you had asked. I have few questions:

> Please add Wandbhook to https://github.com/open-mmlab/mmdetection/pull/7810 once it is complete.
Do you want me to wait before this PR is merged to add the documentation? 

> If possible, please provide a new configuration configs/mask_rcnn/mask_rcnn_r50_fpn_1x_wandb_coco.py as an example, then run this configuration. We will use the link you post as a promotional visualization. Thank you.

What do you mean by link here? Do you want the link to the Weights and Biases project page?

ayulockin commented:
Hey @hhaAndroid I have added the wandb config file as you asked. Can you give it a look and let me know if it's correct.

ayulockin commented:
Hey @hhaAndroid, I have added `after_train_iter` method as well to cover `IterBasedRunner` as well. I hope it helps. I have also cleaned up the code base a bit. Can you let me know what else is needed? :)

hhaAndroid commented:
> Hey @hhaAndroid I have made the changes that you had asked. I have few questions:
> 
> > Please add Wandbhook to #7810 once it is complete.
> 
> Do you want me to wait before this PR is merged to add the documentation?

unnecessary

hhaAndroid commented on configs/mask_rcnn/mask_rcnn_r50_fpn_1x_wandb_coco.py:
```diff
@@ -0,0 +1,20 @@
+                'project': 'mmdetection',
+                'group': 'maskrcnn-r50-fpn-1x-coco'
+             },
+             interval=10,
```
interval=50,
num_eval_images=100

hhaAndroid commented:
> 

Thank you very much, I think this PR is almost enough.

hhaAndroid commented:
> Hey @hhaAndroid, I have added `after_train_iter` method as well to cover `IterBasedRunner` as well. I hope it helps. I have also cleaned up the code base a bit. Can you let me know what else is needed? :)

Do you have resources for training and then release the wandb log?

julled commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,537 @@
+        init_kwargs (dict): A dict passed to wandb.init to initialize
+            a W&B run. Please refer to https://docs.wandb.ai/ref/python/init
+            for possible key-value pairs.
+        interval (int): Logging interval (every k iterations). Defaults to 10.
```
please set default value in the help text as your used value in line 92

julled commented on mmdet/core/hook/wandblogger_hook.py:
```diff
@@ -0,0 +1,537 @@
+            Defaults to True.
+        num_eval_images (int): The number of validation images to be logged.
+            If zero, the evaluation won't be logged. Defaults to 100.
+        bbox_score_thr (float): Threshold for bounding box scores.
```
please add default value for text.

HJoonKwon commented:
@ayulockin Thank you so much!! This PR was very useful for my project, and saved a lot of time. Actually, somehow ```after_train_iter``` was blocking the training/logging process so I just commented it out, and everything works fine. Hope it'd be useful information for you.

ayulockin commented:
Hey @HJoonKwon glad you found it useful. I am curious to know what kind of blocker did you face with `after_train_iter` method present in the code. Can you share any error message or log? Would love to debug it since I am not facing the same on my end.

ayulockin commented:
> Do you have resources for training and then release the wandb log?

Hey @hhaAndroid I do have a script to train using `IterBasedRunner`. I am trying to add a config file named `mask_rcnn_r50_fpn_1x_wandb_iter_coco.py` to train using `IterBasedRunner` however, I am not able to make it run successfully using `python tools/train.py configs/mask_rcnn/mask_rcnn_r50_fpn_1x_wandb_iter_coco.py`. 

It gives this error: `ValueError: IterBasedRunner: Only one of `max_epochs` or `max_iters` can be set.`

Can you help here? Is there any config in MMDetection that uses `IterBasedRunner` that I can take a look at. 

Here's the config I am using but not able to make it work:

```
_base_ = [
    '../_base_/models/mask_rcnn_r50_fpn.py',
    '../_base_/datasets/coco_instance.py',
    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
]

# Setup the runner
runner = dict(type='IterBasedRunner', max_iters=100)
# Set evaluation interval
evaluation = dict(interval=20)
# Set checkpoint interval
checkpoint_config = dict(interval=20)

# yapf:disable
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook'),
        dict(type='MMDetWandbHook',
             init_kwargs={
                'project': 'mmdetection',
                'group': 'maskrcnn-r50-fpn-1x-coco'
             },
             interval=50,
             log_checkpoint=True,
             num_eval_images=100)
        ])
```

hhaAndroid commented:
@ayulockin Hello! Do you have any progress?

ayulockin commented:
> Do you have resources for training and then release the wandb log?

Hey do you need config for training using `IterBasedRunner`?

ayulockin commented:
> @ayulockin Hello! Do you have any progress?

Hey @hhaAndroid I was able to add the config saving feature the way you suggested. I was also able to log the config as W&B config and it feels like a good feature enhancement for the logger. Please let me know if it looks good or now.

I will also provide a colab notebook that can act as a training resource.

hhaAndroid commented:
> colab notebook

it's very good.

hhaAndroid commented on configs/mask_rcnn/mask_rcnn_r50_fpn_1x_wandb_iter_coco.py:
```diff
@@ -0,0 +1,34 @@
+_base_ = [
```
This configuration does not require

ayulockin replied:
I will remove it.

ayulockin commented:
Hey @hhaAndroid here's the colab notebook for training using `MMDetWandbLogger`: https://github.com/ayulockin/myDLexperiments/blob/master/Train_an_Object_Detection%2BSemantic_Segmentation_Model_with_MMDetection_and_W%26B.ipynb

Please let me know if you face any issue while running it. 

I have shared it from my GitHub repo. If you find it useful and educational for community, I can add it to README or docs.

ayulockin commented:
Thanks for your kind comment @ZwwWayne. I will provide a drive link soon.

ayulockin commented:
Hey @ZwwWayne here's the colab link: https://colab.research.google.com/drive/1RCSXHZwDZvakFh3eo9RuNrJbCGqD0dru?usp=sharing

saulgoldsaga commented:
I had to add 
`if not any(masks):
     masks = None`
after line 350 in wandblogger_hook.py
to ensure wandb didn't think I had any masks.

saulgoldsaga commented:
Thanks @ayulockin for putting this together, it has saved me lots of time!
I was wondering if you had done any work on uploading training and validation image examples to wandb?
I thought I should ask before doing myself

ayulockin commented:
> I was wondering if you had done any work on uploading training and validation image examples to wandb?
Hey @saulgoldsaga, this PR adds this feature. If `num_eval_images` is greater than 0 it will log the `validation` images to W&B.

You can use the same logic for training images.
